File: mqtt-kafka-connect/MqttKafkaConnect.js
const mqtt=require('mqtt');
const {Kafka} = require("kafkajs")


class MqttKafkaConnect {

    constructor(mqttIP, mqttTopic, kafkaIp, kafkaClientId, kafkaTopic){
        console.log(kafkaIp)
        this.mqttTopic = mqttTopic;
        const url = "mqtt://"+ mqttIP +":1883"
        const options = {
            clientId:"mqttClient01",
            username:"muhammad",
            password:"123456",
            clean:true
        }

        this.mqttClient = mqtt.connect(url)
        this.kafka = new Kafka({
            "clientId": kafkaClientId,
            "brokers" :[kafkaIp]
        })
        this.kafkaTopic = kafkaTopic
        this.kafkaProducer = this.kafka.producer();

    }

    run(){

        this.mqttClient.on("connect", async () => {
            console.log("connected to Mosquitto");
            await this.kafkaProducer.connect()
            console.log("Connected to Kafka")
            this.mqttClient.subscribe(this.mqttTopic, async (err) => {
                if (!err) {
                    console.log("subscribed to ", this.mqttTopic)
                    // this.mqttClient.publish('mqtttopic', 'Hello mqtt')                    
                }
              })
            // mqttClient.end();
        })
        
        this.mqttClient.on("error", (error) => {
            console.log("Can't connect :: " + error)
            // mqttClient.end();
        })
        
        this.mqttClient.on('message', (topic, message) => {
        
            console.log("Recieved msg: " + message.toString() + "from MQTT on topic " + topic)
            this.sendMsgToKafka(message.toString())
            // mqttClient.end()
        })

    }

    sendMsgToKafka(msg){

        try{
            const result = this.kafkaProducer.send({
                "topic": "devices",
                "messages": [
                    {
                        "value": msg,
                        // "partition": partition
                    }
                ]
            })

            console.log(`Send Successfully! ${JSON.stringify(result)}`)
            // await producer.disconnect();
        }
        catch(err){
            console.error(`ERROR:  ${err}`)
        }
    }
}


module.exports = MqttKafkaConnect

File: mqtt_kafka_connect/mqtt_kafka_connect.py
import paho.mqtt.client as mqtt
from kafka import KafkaProducer
import time


class MqttKafkaConnect:

    def __init__(self, mqtt_ip, mqtt_topic, kafka_ip, kafka_client_id, kafka_topic):

        self.mqtt_client = self._connect_mqtt(mqtt_ip)
        self.mqtt_topic = mqtt_topic
        self.kafka_topic = kafka_topic
        self.kafka_producer = KafkaProducer(bootstrap_servers=kafka_ip)


    def _connect_mqtt(self, broker_ip):
        print(broker_ip)
        client = mqtt.Client("mqttClient01")
        client.connect(broker_ip)
        print("connected to MQTT")
        return client
        

    def run(self):

        def on_message(client, userdata, message):
            print("message recieved from MQTT: " + message.payload.decode("utf-8"))
            self.kafka_producer.send(self.kafka_topic, message.payload.decode("utf-8"))
            print("message sent succefully to Kafka")

        self.mqtt_client.subscribe(self.mqtt_topic)
        self.mqtt_client.on_message = on_message
        self.mqtt_client.loop_forever()
        
